pkg/webhook/v1alpha2/applicationconfiguration/helper.go

```diff
@@ -16,32 +15,102 @@ import (
 )
 
-// checkComponentVersionEnabled check whethter a component is versioning mechanism enabled
-func checkComponentVersionEnabled(ctx context.Context, client client.Reader, dm discoverymapper.DiscoveryMapper,
-	acc *v1alpha2.ApplicationConfigurationComponent) (bool, error) {
-	if acc.RevisionName != "" {
-		return true, nil
-	}
-	for _, ct := range acc.Traits {
-		ut := &unstructured.Unstructured{}
-		if err := json.Unmarshal(ct.Trait.Raw, ut); err != nil {
-			return false, errors.Wrap(err, errUnmarshalTrait)
+// ValidatingAppConfig is used for validating ApplicationConfiguration
+type ValidatingAppConfig struct {
+	appConfig       v1alpha2.ApplicationConfiguration
+	validatingComps []ValidatingComponent
+}
+
+// ValidatingComponent is used for validatiing ApplicationConfigurationComponent
+type ValidatingComponent struct {
+	appConfigComponent v1alpha2.ApplicationConfigurationComponent
+
+	// below data is convenient for validation
+	compName           string
+	component          v1alpha2.Component
+	workloadDefinition v1alpha2.WorkloadDefinition
+	workloadContent    unstructured.Unstructured
+	validatingTraits   []ValidatingTrait
+}
+
+// ValidatingTrait is used for validating Trait
+type ValidatingTrait struct {
+	componentTrait v1alpha2.ComponentTrait
+
+	// below data is convenient for validation
+	traitDefinition v1alpha2.TraitDefinition
+	traitContent    unstructured.Unstructured
+}
+
+// PrepareForValidation prepares data for validations to avoiding repetitive GET/unmarshal operations
+func (v *ValidatingAppConfig) PrepareForValidation(ctx context.Context, c client.Reader, dm discoverymapper.DiscoveryMapper, ac *v1alpha2.ApplicationConfiguration) []error {
+	var allErrors []error
+	v.appConfig = *ac
+	v.validatingComps = make([]ValidatingComponent, len(ac.Spec.Components))
+	for _, acc := range ac.Spec.Components {
+		tmp := ValidatingComponent{}
+		tmp.appConfigComponent = acc
+
+		if acc.ComponentName != "" {
+			tmp.compName = acc.ComponentName
+		} else {
+			tmp.compName = acc.RevisionName
+		}
+		comp, _, err := util.GetComponent(ctx, c, acc, ac.Namespace)
+		if err != nil {
+			allErrors = append(allErrors, errors.Wrapf(err,
+				errFmtGetComponent, tmp.compName))
+			continue
 		}
-		td, err := util.FetchTraitDefinition(ctx, client, dm, ut)
-		if err != nil && !apierrors.IsNotFound(err) {
-			return false, errors.Wrapf(err, errFmtGetTraitDefinition, ut.GetAPIVersion(), ut.GetKind(), ut.GetName())
+		tmp.component = *comp
+
+		// get worload content from raw
+		wl := unstructured.Unstructured{}
+		if err := json.Unmarshal(comp.Spec.Workload.Raw, &wl); err != nil {
+			allErrors = append(allErrors, errors.Wrapf(err,
+				errFmtUnmarshalWorkload, tmp.compName))
+		}
+		tmp.workloadContent = wl
+
+		// get workload definition
+		wlDef, err := util.FetchWorkloadDefinition(ctx, c, dm, &wl)
+		if err != nil {
+			allErrors = append(allErrors, errors.Wrapf(err,
```

Comment:
we don't return all errors at once, please directly return this error.

Reply:
got it.